README.md
# Hide Broken Shortcodes

A plugin for WordPress that prevents broken shortcodes from appearing in posts and pages.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/hide-broken-shortcodes/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress.
3. Optionally filter `hide_broken_shortcode` or `hide_broken_shortcodes_filters` if you want to customize the behavior of the plugin.


## Additional Documentation

* See [readme.txt](https://github.com/coffee2code/hide-broken-shortcodes/blob/master/readme.txt) for additional usage information.
* See [DEVELOPER-DOCS.md](DEVELOPER-DOCS.md) for developer-related documentation on hooks.
* See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/hide-broken-shortcodes/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/hide-broken-shortcodes/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/hide-broken-shortcodes/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 1.9.4 _(2021-10-09)_
* New: Add DEVELOPER-DOCS.md and move hooks documentation into it
* Change: Note compatibility through WP 5.8+
* Change: Tweak installation instruction
* Change: Tweak TODO entry
* Unit tests:
    * Change: Restructure unit test directories
        * Change: Move `phpunit/` into `tests/`
        * Change: Move `phpunit/bin` into `tests/`
    * Change: Remove 'test-' prefix from unit test file
    * Change: In bootstrap, store path to plugin file constant
    * Change: In bootstrap, add backcompat for PHPUnit pre-v6.0

## 1.9.3 _(2021-04-18)_
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 1.9.2 _(2020-09-06)_
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.5+

## 1.9.1 _(2020-08-04)_
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add items to it)
* Change: Add inline documentation for hooks
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Tweak inline documentation formatting
* Unit tests:
    * New: Add test for `get_shortcode_regex()`
    * New: Add test for class name
    * Change: Remove unnecessary unregistering of hooks

## 1.9 _(2019-12-09)_
* New: Add support for shortcodes with names as short as only one character in length (previous minimum was three characters)
* Change: Initialize plugin on `plugins_loaded` action instead of on load
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* Unit tests:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * New: Add additional test data that includes shortcodes using single quotes around their attribute values
    * Fix: Prevent theoretical warning about undefined variable
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)
* Change: Update License URI to be HTTPS
* Change: Split paragraph in README.md's "Support" section into two

## 1.8.2 _(2018-06-29)_
* New: Bail early if text doesn't contain a square bracket (and thus no shortcodes)
* New: Add README.md
* New: Add unit tests for square brackets in HTML comments
* New: Add GitHub link to readme
* Change: Minor whitespace tweaks to unit test bootstrap
* Change: Add item to FAQ
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Note compatibility through WP 4.9+
* Change: Update copyright date (2018)

## 1.8.1 _(2017-02-08)_
* Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable.
* Change: Enable more error output for unit tests.
* Change: Add more unit tests.
* Change: Note compatibility through WP 4.7+.
* Change: Miscellaneous readme.txt improvements.
* Change: Minor code documentation reformatting.
* Change: Update copyright date (2017).
* New: Add LICENSE file.

## 1.8 _(2016-05-21)_
* Bugfix: Don't attempt to hide shortcodes (or what may look like shortcodes) appearing within HTML tags.
* New: Add unit test to ensure shortcode escape notation is not hidden by the plugin.
* Change: Prevent web invocation of unit test bootstrap.php.
* Change: Note compatibility through WP 4.5+.

## 1.7.1 _(2016-01-27)_
* Change: Register hooks during `plugins_loaded` instead of `init`.
* New: Add support for language packs:
    * Define 'Text Domain' header attribute.
    * Load textdomain.
* New: Create empty index.php to prevent files from being listed if web server has enabled directory listings.
* Change: Note compatibility through WP 4.4+.
* Change: Explicitly declare methods in unit tests as public.
* Change: Update copyright date (2016).

## 1.7 _(2015-04-02)_
* Enhancement: Filter `the_excerpt` by default as well
* Update: Add more unit tests
* Update: Note compatibility through WP 4.2+
* Update: Add inline documentation to examples in readme.txt
* Update: Minor inline documentation tweaks (spacing, formatting)

## 1.6.3 _(2015-02-14)_
* Add trivial unit test for plugin version
* Note compatibility through WP 4.1+
* Update copyright date (2015)

## 1.6.2 _(2014-08-30)_
* Minor plugin header reformatting
* Minor code reformatting (bracing)
* Change documentation links to wp.org to be https
* Note compatibility through WP 4.0+
* Add plugin icon

## 1.6.1 _(2013-12-29)_
* Add unit tests
* Note compatibility through WP 3.8+
* Update copyright date (2014)
* Minor readme.txt tweaks
* Change donate link
* Add banner

## 1.6
* Update regex to allow hyphens in shortcode names (syncing changes made in WP 3.5)
* Add check to prevent execution of code if file is directly accessed
* Note compatibility through WP 3.5+
* Update copyright date (2013)

## 1.5
* Recursively hide nested broken shortcodes
* Re-license as GPLv2 or later (from X11)
* Add 'License' and 'License URI' header tags to readme.txt and plugin file
* Remove ending PHP close tag
* Note compatibility through WP 3.4+
* Fix error in example code in readme.txt

## 1.4
* Update `get_shortcode_regex()` and `do_shortcode_tag()` to support shortcode escape syntax
* NOTE: The preg match array sent via the `hide_broken_shortcode` filter has changed and requires you to update any code that hooks it
* Add `version()` to return plugin version
* Note compatibility through WP 3.3+
* Add Frequently Asked Questions section to readme.txt
* Add link to plugin directory page to readme.txt
* Update copyright date (2012)

## 1.3.1

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Hide_Broken_Shortcodes
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/hide-broken-shortcodes.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Hide_Broken_Shortcodes
 */

define( 'HIDE_BROKEN_SHORTCODES_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/hide-broken-shortcodes.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require HIDE_BROKEN_SHORTCODES_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-hide-broken-shortcodes.php
<?php

defined( 'ABSPATH' ) or die();

class Hide_Broken_Shortcodes_Test extends WP_UnitTestCase {

	public function setUp() {
		parent::setUp();

		add_shortcode( 'abcabc', array( $this, 'handle_shortcode' ) );
		add_shortcode( 'b',      array( $this, 'handle_shortcode' ) );
	}

	public function tearDown() {
		parent::tearDown();

		remove_shortcode( 'abcabc' );
		remove_shortcode( 'b' );
	}


	//
	//
	// DATA PROVIDERS
	//
	//


	public function text_shortcode_without_content() {
		return array(
			array( 'This contains [unknown] shortcode.' ),
			array( 'This contains [unknown/] shortcode.' ),
			array( 'This contains [unknown aaa="dog"] shortcode.'),
			array( 'This contains [unknown aaa="dog" /] shortcode.'),
			array( 'This contains [unknown aaa=\'dog\' /] shortcode.'),
			array( 'This contains [unknown][/unknown] shortcode.'),
			array( 'This contains [tp][/tp] shortcode.'),
			array( 'This contains [i][/i] shortcode.'),
		);
	}

	public function text_shortcode_with_content() {
		return array(
			array( 'This contains [unknown]abc[/unknown] shortcode.'),
			array( 'This contains [unknown aaa="emu"]abc[/unknown] shortcode.'),
			array( 'This contains [unknown aaa=\'emu\']abc[/unknown] shortcode.'),
			array( 'This contains [tp]abc[/tp] shortcode.'),
			array( 'This contains [i]abc[/i] shortcode.'),
		);
	}

	public function default_filters() {
		return array(
			array( 'the_content' ),
			array( 'the_excerpt' ),
			array( 'widget_text' ),
		);
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	public function handle_shortcode( $atts, $content = null ) {
		$defaults = array(
			'aaa' => 'cat',
			'bbb' => '',
		);
		$atts = shortcode_atts( $defaults, $atts );

		$ret = '(' . $atts['aaa'];
		if ( $atts['bbb'] )
			$ret .= "," . $atts['bbb'];
		$ret .= ')';

		if ( $content )
			$ret .= $content . '(end)';

		return $ret;
	}

	public function prevent_shortcode_hiding( $default_display, $shortcode_name, $match_array ) {
		$shortcodes_not_to_hide = array( 'i', 'tp', 'unknown' );
		$display = '';
		if ( in_array( $shortcode_name, $shortcodes_not_to_hide ) ) {
			$display = $match_array[0];
		}
		return $display;
	}

	public function filter_the_title( $filters ) {
		$filters[] = 'the_title';
		return $filters;
	}


	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_HideBrokenShortcodes' ) );
	}

	public function test_version() {
		$this->assertEquals( '1.9.1', c2c_HideBrokenShortcodes::version() );
	}

	/**
	 * @dataProvider default_filters
	 */
	public function test_hooks_default_filter( $filter ) {
		$this->assertEquals( 1001, has_filter( $filter, array( 'c2c_HideBrokenShortcodes', 'do_shortcode' ) ) );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_HideBrokenShortcodes', 'init' ) ) );
	}


	/**
	 * @dataProvider text_shortcode_without_content
	 */
	public function test_hooks_custom_filter( $text ) {
		$this->assertEquals( apply_filters( 'the_title', $text ), apply_filters( 'the_title', $text ) );

		add_filter( 'hide_broken_shortcodes_filters', array( $this, 'filter_the_title' ) );
		c2c_HideBrokenShortcodes::register_filters(); // Pretend the filter was added before plugin initialized.

		$this->assertEquals( 'This contains  shortcode.', apply_filters( 'the_title', $text ) );
	}

	public function test_handled_shortcode_not_affected() {
		$this->assertEquals( 'hello (fox)', do_shortcode( 'hello [abcabc aaa="fox"]' ) );
		$this->assertEquals( '(fox)', trim( apply_filters( 'the_content', '[abcabc aaa="fox"]' ) ) );
		$this->assertEquals( '(hippo)hiphop(end)', trim( apply_filters( 'the_content', '[abcabc aaa="hippo"]hiphop[/abcabc]' ) ) );

		$this->assertEquals( 'hello (fox)', do_shortcode( 'hello [b aaa="fox"]' ) );
		$this->assertEquals( '(fox)', trim( apply_filters( 'the_content', '[b aaa="fox"]' ) ) );
		$this->assertEquals( '(hippo)hiphop(end)', trim( apply_filters( 'the_content', '[b aaa="hippo"]hiphop[/b]' ) ) );
	}

	public function test_shortcode_escape_notation_not_affected() {
		$text     = 'Use [[shortcode]] like so.';
		$expected = 'Use [shortcode] like so.';

		$this->assertEquals( wpautop( $expected ), apply_filters( 'the_content', $text ) );
	}

	public function test_does_not_affect_bracket_usage_within_html_tags() {
		$html = array(
			'<input type="text" name="hello[world]" />',
			'<input type="text" name="hello[\'world\']" />',
			"<input type='text' name='hello[\"world\"]' />",
			"<span class='example' title='What if cart[qty] were here?'>Test</span>",
			'<span class="example" title="What if cart[\'qty\'] were here?">Test</span>',
			'<span class="example" title="What if cart[\"qty\"] were here?">Test</span>',
			'<span class="example" title="What if cart[qty name="yes"] were here?">Test</span>',
			"If this <cat[qty]>.",
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( $h ), apply_filters( 'the_content', $h ) );
		}
	}

	public function test_does_not_affect_bracketed_integers() {
		$html = array(
			'See [1] for more info.',
			'There are [45] here.',
			'There are [ 16 ] here.',
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( $h ), apply_filters( 'the_content', $h ) );
		}
	}

	public function test_does_not_affect_bracketed_quoted_strings() {
		$html = array(
			'Code example of $cat["name"]',
			'Code example of $cat[ "name" ]',
			"Give it a name of cart[{\$cart_item_key}]['qty']",
			"Give it a name of cart[ {\$cart_item_key} ][ 'qty' ]",
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( wptexturize( $h ) ), apply_filters( 'the_content', $h ) );
		}
	}

	public function test_does_not_affect_bracketed_variable() {
		$html = array(
			'Code example of [$cat].',
			'Give it a name of $cart[ $cat_food ]',
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( $h ), apply_filters( 'the_content', $h ) );
		}
	}

	public function test_does_not_affect_empty_brackets() {
		$html = array(
			'Code example of [].',
			"Give it a name of cart[{\$cart_item_key}][]",
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( $h ), apply_filters( 'the_content', $h ) );
		}
	}

	public function test_does_not_affect_html_comments() {
		$html = array(
			'<![CDATA[Some text]]>',
			'<!-- This is [some] text -->',
		);

		foreach ( $html as $h ) {
			$this->assertEquals( wpautop( $h ), apply_filters( 'the_content', $h ) );
		}
	}

	/**
	 * @dataProvider text_shortcode_without_content
	 */
	public function test_unhandled_shortcodes_without_content_get_hidden( $text ) {
		$this->assertEquals( 'This contains  shortcode.', apply_filters( 'widget_text', $text ) );
	}

	/**
	 * @dataProvider text_shortcode_with_content
	 */
	public function test_unhandled_shortcodes_witb_content_get_hidden_but_content_remains( $text ) {
		$this->assertEquals( 'This contains abc shortcode.', apply_filters( 'widget_text', $text ) );
	}

	/**
	 * @dataProvider text_shortcode_without_content
	 */
	public function test_filter_to_prevent_unhandled_shortcodes_from_getting_hidden( $text ) {
		add_filter( 'hide_broken_shortcode', array( $this, 'prevent_shortcode_hiding' ), 10, 3 );

		$this->assertEquals( $text, apply_filters( 'widget_text', $text ) );
	}

	/**
	 * @dataProvider text_shortcode_with_content
	 */
	public function test_filter_to_prevent_unhandled_shortcodes_with_content_from_getting_hidden( $text ) {
		add_filter( 'hide_broken_shortcode', array( $this, 'prevent_shortcode_hiding' ), 10, 3 );

		$this->assertEquals( $text, apply_filters( 'widget_text', $text ) );
	}

	/*
	 * get_shortcode_regex()
	 */

	// This is primarily just to test that the regex isn't unintentionally tweaked
	// in such a way that isn't caught by other tests.
	public function test_get_shortcode_regex() {
		$expected = '(?!<.*)\[(\[?)([a-zA-Z_\-][0-9a-zA-Z_\-\+]*)(?![\w-])([^\]\/]*(?:\/(?!\])[^\]\/]*)*?)(?:(\/)\]|\](?:([^\[]*+(?:\[(?!\/\2\])[^\[]*+)*+)\[\/\2\])?)(\]?)(?![^<>]*>)';

		$this->assertEquals( $expected, c2c_HideBrokenShortcodes::get_shortcode_regex() );
	}
}


hide-broken-shortcodes.php
<?php
/**
 * Plugin Name: Hide Broken Shortcodes
 * Version:     1.9.4
 * Plugin URI:  https://coffee2code.com/wp-plugins/hide-broken-shortcodes/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: hide-broken-shortcodes
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Prevent broken shortcodes from appearing in posts and pages.
 *
 * Compatible with WordPress 2.5 through 5.8+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/hide-broken-shortcodes/
 *
 * @package Hide_Broken_Shortcodes
 * @author  Scott Reilly
 * @version 1.9.4
 */

/*
	Copyright (c) 2009-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_HideBrokenShortcodes' ) ) :

class c2c_HideBrokenShortcodes {

	/**
	 * Returns version of the plugin.
	 *
	 * @since 1.4
	 */
	public static function version() {
		return '1.9.4';
	}

	/**
	 * Initializes class variables and adds actions and filters.
	 */
	public static function init() {
		self::register_filters();
	}

	/**
	 * Registers filters and loads text domain.
	 *
	 * @since 1.7 Filter 'the_excerpt' by default as well.
	 */
	public static function register_filters() {
		// Load plugin textdomain.
		load_plugin_textdomain( 'hide-broken-shortcodes' );

		/**
		 * Filters the list of filters that get processed for broken shortcode hiding.
		 *
		 * @param array $filters The list of filters. Default ['the_content', 'the_excerpt', 'widget_text'].
		 */
		$filters = (array) apply_filters( 'hide_broken_shortcodes_filters', array( 'the_content', 'the_excerpt', 'widget_text' ) );

		foreach ( $filters as $filter ) {
			add_filter( $filter, array( __CLASS__, 'do_shortcode' ), 1001 ); // Do this after the built-in do_shortcode() operates, which is 11
		}
	}

	/**
	 * Like WP's `do_shortcode()`, but doesn't return content immediately if no
	 * shortcodes exist.
	 *
	 * @param  string $content The primary text to be processed for shortcodes.
	 * @return string
	 */
	public static function do_shortcode( $content ) {
		if ( false === strpos( $content, '[' ) ) {
			return $content;
		}

		return preg_replace_callback( '/' . self::get_shortcode_regex() . '/s', array( __CLASS__, 'do_shortcode_tag' ), $content );
	}

	/**
	 * Like WP's `get_shortcode_regex()`, but matches for anything that looks
	 * like a shortcode.
	 *
	 * @return string The regexp for finding shortcodes in text.
	 */
	public static function get_shortcode_regex() {
		$tagregexp = '[a-zA-Z_\-][0-9a-zA-Z_\-\+]*';

		// WARNING! Do not change this regex without changing do_shortcode_tag()
		return '(?!<.*)'                         // Non-capturing check that text within HTML tags are skipped
			. '\\['                              // Opening bracket
			. '(\\[?)'                           // 1: Optional second opening bracket for escaping shortcodes: [[tag]]
			. "($tagregexp)"                     // 2: Shortcode name
			. '(?![\\w-])'                       // Not followed by word character or hyphen
			. '('                                // 3: Unroll the loop: Inside the opening shortcode tag
			.     '[^\\]\\/]*'                   // Not a closing bracket or forward slash
			.     '(?:'
			.         '\\/(?!\\])'               // A forward slash not followed by a closing bracket
			.         '[^\\]\\/]*'               // Not a closing bracket or forward slash
			.     ')*?'
			. ')'
			. '(?:'
			.     '(\\/)'                        // 4: Self closing tag ...
			.     '\\]'                          // ... and closing bracket
			. '|'
			.     '\\]'                          // Closing bracket
			.     '(?:'
			.         '('                        // 5: Unroll the loop: Optionally, anything between the opening and closing shortcode tags
			.             '[^\\[]*+'             // Not an opening bracket
			.             '(?:'
			.                 '\\[(?!\\/\\2\\])' // An opening bracket not followed by the closing shortcode tag
			.                 '[^\\[]*+'         // Not an opening bracket
			.             ')*+'
			.         ')'
			.         '\\[\\/\\2\\]'             // Closing shortcode tag
			.     ')?'
			. ')'
			. '(\\]?)'                           // 6: Optional second closing brocket for escaping shortcodes: [[tag]]
			. '(?![^<>]*>)';                     // Non-capturing check that text within HTML tags are skipped
	}

	/**
	 * Callback to handle each shortcode not replaced via the traditional
	 * shortcode system.
	 *
	 * The actual replacement string used can be modified by filtering
	 * 'hide_broken_shortcode'. By default this is the text between the
	 * opening/closing shortcode tags, or an empty string if there was no
	 * closing tag.
	 *
	 * @param  string $m The preg_match result array for the unhandled shortcode.
	 * @return string    The replacement string for the unhandled shortcode.
	 */
	public static function do_shortcode_tag( $m ) {
		// If this function gets executed, then the shortcode found is not being handled.

		// allow [[foo]] syntax for escaping a tag
		if ( $m[1] == '[' && $m[6] == ']' ) {
			return substr( $m[0], 1, -1 );
		}

		// If text is being wrapped by opening and closing shortcode tag, show text. Otherwise, show nothing.
		$default_display = ( isset( $m[5] ) ? self::do_shortcode( $m[5] ) : '' );

		/**
		 * Filters the list of filters that get processed for broken shortcode hiding.
		 *
		 * @param array  $default_display The text within opening and closing shortcode tags, else an empty string.
		 * @param string $shortcode_name  The name of the shortcode.
		 * @param array  $match_array     The preg_match result array for the unhandled shortcode.
		 */
		return apply_filters( 'hide_broken_shortcode', $default_display, $m[2], $m );
	}

} // end c2c_HideBrokenShortcodes

add_action( 'plugins_loaded', array( 'c2c_HideBrokenShortcodes', 'init' ) );

endif; // end if !class_exists()
